MdeModulePkg/Library/MemoryBucketLib/MemoryBucketLib.inf

```diff
@@ -0,0 +1,36 @@
+## @file
+# Library the defines Runtime memory buckets for use during PEI
+#
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+##
+
+[Defines]
+  INF_VERSION                    = 0x00010005
+  BASE_NAME                      = MemoryBucketLib
+  FILE_GUID                      = 916b6cbe-52f3-439b-8f4c-b97453e36cc7
+  MODULE_TYPE                    = PEIM
+  VERSION_STRING                 = 1.0
+  LIBRARY_CLASS                  = MemoryBucketLib|DXE_CORE DXE_DRIVER DXE_RUNTIME_DRIVER DXE_SMM_DRIVER SMM_CORE PEIM SEC PEI_CORE UEFI_APPLICATION UEFI_DRIVER

```

Comment:
You probably need two separate libraries. PEI has a lot of functionality specific to PEI (producing HOB info / checking info for allocations). DXE is consumer focused. So, either library interfaces are exposed to DXE that don't make sense and/or a lot of irrelevant functionality is being linked to DXE.

In either case, if something is only intended in PEI, place it in a PEI (even `PEI_CORE`) specific library interface/instance. Otherwise, the DXE type that makes sense.

Reply:
There's also logic present like HOB producing logic that can't be used in DXE.

Reply:
Switched to just PEI_CORE.  This shouldn't be used by anything else.